File: Sources/ComicCore/EPUBParser.swift
import Foundation
import SwiftSoup

private final class PackageXML: NSObject, XMLParserDelegate {
    var rootfiles: [String] = []
    var manifest: [String: [String: String]] = [:]
    var spine: [[String: String]] = []
    var title = ""
    var direction: ReadingDirection?
    var coverID: String?
    var element = ""
    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes: [String: String]) {
        element = elementName.components(separatedBy: ":").last ?? elementName
        switch element {
        case "rootfile": if let path = attributes["full-path"] { rootfiles.append(path) }
        case "item": if let id = attributes["id"] { manifest[id] = attributes }
        case "itemref": spine.append(attributes)
        case "spine": direction = attributes["page-progression-direction"].flatMap(ReadingDirection.init(rawValue:))
        case "meta": if attributes["name"] == "cover" { coverID = attributes["content"] }
        default: break
        }
    }
    func parser(_ parser: XMLParser, foundCharacters string: String) { if element == "title" { title += string } }
    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) { element = "" }
    static func read(_ data: Data) throws -> PackageXML {
        let delegate = PackageXML(); let parser = XMLParser(data: data)
        parser.shouldResolveExternalEntities = false; parser.delegate = delegate
        guard parser.parse() else { throw ReaderError.message("EPUB 结构 XML 无法解析：\(parser.parserError?.localizedDescription ?? "未知错误")") }
        return delegate
    }
}

public enum EPUBParser {
    public static func parse(archive: BookArchive) throws -> (String, [ReadingUnit], ReadingDirection?, [NavigationItem], [String]) {
        let container = try PackageXML.read(archive.data("META-INF/container.xml", limit: 1024 * 1024))
        guard let packagePath = container.rootfiles.first else { throw ReaderError.message("EPUB 缺少包文档入口。") }
        let package = try PackageXML.read(archive.data(try BookArchive.normalize(packagePath), limit: 16 * 1024 * 1024))
        let coverItem = package.coverID.flatMap { package.manifest[$0] } ?? package.manifest.values.first { $0["properties"]?.contains("cover-image") == true }
        let coverImage = try coverItem?["href"].map { try BookArchive.resolve($0, relativeTo: packagePath) }
        var stylesheets: [String: String] = [:]
        for item in package.manifest.values where item["media-type"] == "text/css" {
            if let href = item["href"], let path = try? BookArchive.resolve(href, relativeTo: packagePath), let data = try? archive.data(path, limit: 2 * 1024 * 1024) {
                stylesheets[path] = String(decoding: data, as: UTF8.self)
            }
        }
        var units: [ReadingUnit] = []; var warnings: [String] = []
        for (occurrence, item) in package.spine.enumerated() {
            try Task.checkCancellation()
            if item["linear"] == "no" { continue }
            guard let ref = item["idref"], let manifest = package.manifest[ref], let href = manifest["href"] else {
                units.append(ReadingUnit(locator: SourceLocator(resource: "missing-\(occurrence)", occurrence: occurrence), title: "缺失页面", error: "spine 引用不存在。")); continue
            }
            let path = try BookArchive.resolve(href, relativeTo: packagePath)
            let locator = SourceLocator(resource: path, occurrence: occurrence)
            guard let data = try? archive.data(path, limit: 16 * 1024 * 1024) else {
                units.append(ReadingUnit(locator: locator, title: "第 \(units.count+1) 页", error: "内容文档无法读取。")); continue
            }
            let document = try SwiftSoup.parse(String(decoding: data, as: UTF8.self))
            var css = ""
            for node in try document.select("style, link[rel=stylesheet]").array() {
                let media = try node.attr("media").trimmingCharacters(in:.whitespacesAndNewlines)
                let sheet: String?
                if node.tagName() == "style" { sheet = try node.html() }
                else if let linkedPath = try? BookArchive.resolve(node.attr("href"),relativeTo:path) { sheet=stylesheets[linkedPath] }
                else { sheet=nil }
                if let sheet { css += "\n" + (media.isEmpty ? sheet : "@media \(media) {\(sheet)}") }
            }
            let images = try document.select("body img").array()
            let bodyText = try document.body()?.text() ?? ""
            let hasComplexNodes = try !document.select("svg, canvas, video, audio, iframe, script, object, table").isEmpty()
            var imagePath: String?; var hint: Int?
            let image = images.count == 1 ? images.first : nil
            if let image {
                imagePath = try? BookArchive.resolve(image.attr("src"), relativeTo: path)
                hint = RotationStyles.hint(css:css,image:image)
            }
            let isCover = imagePath != nil && imagePath == coverImage
            let layoutStyles = css + (try document.select("[style]").array().map { try $0.attr("style") }.joined(separator: "\n"))
            let riskyLayout = layoutStyles.range(of: "position\\s*:\\s*(absolute|fixed)|clip(?:-path)?\\s*:|background(?:-image)?\\s*:.*url\\(|@import", options: .regularExpression) != nil
            let complex = imagePath == nil || !bodyText.isEmpty || hasComplexNodes || riskyLayout
            let title = isCover ? "封面" : (try document.title()).trimmingCharacters(in: .whitespacesAndNewlines)
            units.append(ReadingUnit(locator: locator, title: title.isEmpty ? "第 \(units.count+1) 页" : title, imagePath: imagePath, isCover: isCover, rotationHint: hint, complex: complex))
        }
        guard !units.isEmpty else { throw ReaderError.message("这本 EPUB 没有可阅读的正文内容。") }
        var navigation: [NavigationItem] = []
        for manifest in package.manifest.values where manifest["media-type"] == "application/x-dtbncx+xml" || manifest["properties"]?.contains("nav") == true {
            guard let href = manifest["href"], let path = try? BookArchive.resolve(href, relativeTo: packagePath), let data = try? archive.data(path, limit: 4 * 1024 * 1024), let document = try? SwiftSoup.parse(String(decoding: data, as: UTF8.self)) else { continue }
            if manifest["media-type"] == "application/x-dtbncx+xml" {
                for node in try document.select("navpoint").array() {
                    guard let target = try node.select("content").first()?.attr("src"), let resolved = try? BookArchive.resolve(target, relativeTo: path), let index = units.firstIndex(where: { $0.locator.resource == resolved }) else { continue }
                    navigation.append(NavigationItem(title: try node.select("navlabel").text(), index: index))
                }
            } else {
                for node in try document.select("nav a").array() {
                    guard let resolved = try? BookArchive.resolve(node.attr("href"), relativeTo: path), let index = units.firstIndex(where: { $0.locator.resource == resolved }) else { continue }
                    navigation.append(NavigationItem(title: try node.text(), index: index))
                }
            }
        }
        if archive.contains("META-INF/encryption.xml") { warnings.append("出版物含加密或字体混淆声明；受保护的资源可能无法显示。") }
        return (package.title.trimmingCharacters(in: .whitespacesAndNewlines), units, package.direction, navigation, warnings)
    }
}

File: android/app/src/main/java/com/mandu/reader/data/EpubParser.kt
package com.mandu.reader.data

import com.mandu.reader.core.NavigationItem
import com.mandu.reader.core.ReadingDirection
import com.mandu.reader.core.ReadingUnit
import com.mandu.reader.core.SourceLocator
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.parser.Parser

internal data class EpubData(
    val title: String,
    val units: List<ReadingUnit>,
    val direction: ReadingDirection?,
    val navigation: List<NavigationItem>,
    val warnings: List<String>
)

internal object EpubParser {
    private const val XML_LIMIT = 16L * 1024 * 1024
    private const val STYLESHEET_TOTAL_LIMIT = 64L * 1024 * 1024
    private data class ManifestItem(
        val id: String,
        val href: String,
        val mediaType: String,
        val properties: Set<String>
    )

    fun parse(archive: ZipArchive, checkActive: () -> Unit = {}): EpubData {
        val container = xml(archive.bytes("META-INF/container.xml", 1024L * 1024))
        val packageReference = elements(container, "rootfile").firstOrNull()?.attr("full-path")
            ?.takeIf(String::isNotBlank) ?: throw ArchiveException("EPUB 缺少包文档入口。")
        val packagePath = ZipArchive.normalize(packageReference)
        val opf = xml(archive.bytes(packagePath, XML_LIMIT))
        val manifestItems = elements(opf, "item").mapNotNull { element ->
            val id = element.attr("id")
            val href = element.attr("href")
            if (id.isBlank() || href.isBlank()) null else ManifestItem(
                id,
                href,
                element.attr("media-type").lowercase(),
                element.attr("properties").split(Regex("\\s+")).filter(String::isNotBlank).toSet()
            )
        }
        if (manifestItems.map { it.id }.toSet().size != manifestItems.size) {
            throw ArchiveException("EPUB manifest 包含重复 id。")
        }
        val manifest = manifestItems.associateBy { it.id }
        val title = elements(opf, "title").firstOrNull()?.text()?.trim().orEmpty()
        val spineElement = elements(opf, "spine").firstOrNull()
        val direction = when (spineElement?.attr("page-progression-direction")?.lowercase()) {
            "rtl" -> ReadingDirection.RTL
            "ltr" -> ReadingDirection.LTR
            else -> null
        }
        val epub2CoverId = elements(opf, "meta").firstOrNull {
            it.attr("name").equals("cover", true)
        }?.attr("content")
        val coverItem = epub2CoverId?.let(manifest::get)
            ?: manifest.values.firstOrNull { "cover-image" in it.properties }
        val coverImagePath = coverItem?.let { runCatching { ZipArchive.resolve(it.href, packagePath) }.getOrNull() }

        val stylesheetText = mutableMapOf<String, String>()
        var stylesheetBytes = 0L
        for (item in manifest.values.filter { it.mediaType == "text/css" }) {
            checkActive()
            val path = runCatching { ZipArchive.resolve(item.href, packagePath) }.getOrNull() ?: continue
            val bytes = runCatching { archive.bytes(path, 2L * 1024 * 1024) }.getOrNull() ?: continue
            stylesheetBytes += bytes.size
            if (stylesheetBytes > STYLESHEET_TOTAL_LIMIT) throw ArchiveException("EPUB 样式资源超过 64 MiB 解析预算。")
            stylesheetText[path] = bytes.toString(Charsets.UTF_8)
        }
        fun stylesheet(path: String): String? {
            stylesheetText[path]?.let { return it }
            val bytes = try { archive.bytes(path, 2L * 1024 * 1024) } catch (_: Exception) { return null }
            stylesheetBytes += bytes.size
            if (stylesheetBytes > STYLESHEET_TOTAL_LIMIT) throw ArchiveException("EPUB 样式资源超过 64 MiB 解析预算。")
            return bytes.toString(Charsets.UTF_8).also { stylesheetText[path] = it }
        }

        val units = mutableListOf<ReadingUnit>()
        elements(opf, "itemref").forEachIndexed { occurrence, itemRef ->
            checkActive()
            if (itemRef.attr("linear").equals("no", true)) return@forEachIndexed
            val idref = itemRef.attr("idref")
            val item = manifest[idref]
            if (item == null) {
                units += ReadingUnit(
                    SourceLocator("missing-spine-$occurrence", occurrence),
                    "缺失页面",
                    error = "spine 引用不存在：${idref.ifBlank { "(空)" }}"
                )
                return@forEachIndexed
            }
            val contentPath = runCatching { ZipArchive.resolve(item.href, packagePath) }.getOrElse {
                units += ReadingUnit(
                    SourceLocator("invalid-spine-$occurrence", occurrence),
                    "缺失页面",
                    error = "正文资源路径无效。"
                )
                return@forEachIndexed
            }
            val locator = SourceLocator(contentPath, occurrence)
            val bytes = runCatching { archive.bytes(contentPath, XML_LIMIT) }.getOrElse {
                units += ReadingUnit(locator, "第 ${units.size + 1} 页", error = "内容文档无法读取。")
                return@forEachIndexed
            }
            val document = html(bytes)
            val css = buildString {
                for (node in document.select("style, link[rel~=(?i)stylesheet]")) {
                    val media = node.attr("media").trim()
                    val sheet = if (node.normalName() == "style") node.data().ifBlank { node.html() }
                    else runCatching { ZipArchive.resolve(node.attr("href"), contentPath) }.getOrNull()?.let(::stylesheet)
                    if (sheet != null) append('\n').append(if (media.isBlank()) sheet else "@media $media {$sheet}")
                }
            }
            val images = document.select("body img[src]")
            val image = images.singleOrNull()
            val imagePath = image?.let { runCatching { ZipArchive.resolve(it.attr("src"), contentPath) }.getOrNull() }
            val hint = image?.let { RotationStyles.hint(css, it) }
            val bodyText = document.body()?.text()?.trim().orEmpty()
            val complexNodes = document.select("svg, canvas, video, audio, iframe, script, object, table, picture").isNotEmpty()
            val inlineStyles = document.select("[style]").joinToString("\n") { it.attr("style") }
            val layoutStyles = "$css\n$inlineStyles"
            val riskyLayout = Regex(
                "position\\s*:\\s*(absolute|fixed)|clip(?:-path)?\\s*:|background(?:-image)?\\s*:.*url\\(|@import",
                RegexOption.IGNORE_CASE
            ).containsMatchIn(layoutStyles)
            val complex = imagePath == null || images.size != 1 || bodyText.isNotEmpty() || complexNodes || riskyLayout
            val isCover = imagePath != null && imagePath == coverImagePath
            val imageBytes = imagePath?.let { path -> runCatching { archive.bytes(path) }.getOrNull() }
            val isTiff = imagePath != null && extension(imagePath) in setOf("tif", "tiff")
            val tiffResult = if (isTiff && imageBytes != null) runCatching {
                TiffDecoder.pages(imageBytes).firstOrNull()
            } else null
            val tiffPage = tiffResult?.getOrNull()
            val dimensions = if (tiffPage != null) ImageInfo(tiffPage.width, tiffPage.height, 1)
            else imageBytes?.let { runCatching { ImageDecoder.info(it) }.getOrNull() }
            val pageTitle = if (isCover) "封面" else document.title().trim().ifBlank { "第 ${units.size + 1} 页" }
            units += ReadingUnit(
                locator = locator,
                title = pageTitle,
                imagePath = imagePath,
                width = dimensions?.width ?: 0,
                height = dimensions?.height ?: 0,
                isCover = isCover,
                rotationHint = hint,
                complex = complex,
                error = when {
                    image != null && imagePath == null -> "页面图片路径无效。"
                    imagePath != null && imageBytes == null -> "页面图片无法读取。"
                    isTiff && tiffPage == null -> tiffResult?.exceptionOrNull()?.message ?: "TIFF 页面无法读取。"
                    tiffPage?.error != null -> tiffPage.error
                    else -> null
                }
            )
        }
        if (units.isEmpty()) throw ArchiveException("这本 EPUB 没有可阅读的正文内容。")

        val navigation = mutableListOf<NavigationItem>()
        val tocItems = manifest.values.filter {
            it.mediaType == "application/x-dtbncx+xml" || "nav" in it.properties
        }
        for (item in tocItems) {
            val tocPath = runCatching { ZipArchive.resolve(item.href, packagePath) }.getOrNull() ?: continue
            val tocBytes = runCatching { archive.bytes(tocPath, 4L * 1024 * 1024) }.getOrNull() ?: continue
            val document = if (item.mediaType == "application/x-dtbncx+xml") xml(tocBytes) else html(tocBytes)
            if (item.mediaType == "application/x-dtbncx+xml") {
                for (point in elements(document, "navpoint")) {
                    val target = elements(point, "content").firstOrNull()?.attr("src").orEmpty()
                    addNavigation(navigation, units, target, tocPath, elements(point, "text").firstOrNull()?.text().orEmpty())
                }
            } else {
                val nav = document.select("nav").firstOrNull {
                    it.attr("epub:type").split(Regex("\\s+")).any { token -> token == "toc" }
                } ?: document.selectFirst("nav")
                nav?.select("a[href]")?.forEach { link ->
                    addNavigation(navigation, units, link.attr("href"), tocPath, link.text())
                }
            }
        }
        val warnings = buildList {
            if (archive.contains("META-INF/encryption.xml")) {
                add("出版物含加密或字体混淆声明；受保护的资源可能无法显示。")
            }
        }
        return EpubData(title, units, direction, navigation, warnings)
    }

    private fun addNavigation(
        output: MutableList<NavigationItem>,
        units: List<ReadingUnit>,
        target: String,
        relativeTo: String,
        title: String
    ) {
        val path = runCatching { ZipArchive.resolve(target, relativeTo) }.getOrNull() ?: return
        val index = units.indexOfFirst { it.locator.resource == path }
        if (index >= 0) output += NavigationItem(title.trim().ifBlank { "第 ${index + 1} 页" }, index, ZipArchive.fragment(target))
    }

    private fun xml(bytes: ByteArray): Document =
        Jsoup.parse(bytes.toString(Charsets.UTF_8), "", Parser.xmlParser())

    private fun html(bytes: ByteArray): Document = Jsoup.parse(bytes.toString(Charsets.UTF_8))

    private fun elements(root: Element, localName: String): List<Element> =
        root.getAllElements().filter { it.normalName().substringAfterLast(':').equals(localName, true) }
}
